### ISSUE TITLE
Thoughts on driver wrapping

### ISSUE THREAD

**Issue by pudelkoM** [emmericp/ixy#6]
Universal structure contained in every driver and exposed to the user:
```C
struct ixy_device {
    char* driver_name;
    uint16_t (*tx_batch)(struct ixy_device*, struct pkt_buf*, uint16_t);
    uint16_t (*rx_batch)(struct ixy_device*, struct pkt_buf*, uint16_t);
};
```

Specific driver implementations include this struct somewhere:
```C
struct virtio_device {
    int fd;
    struct virt_queue* rx, *tx, *ctrl;

    struct ixy_device ixy;
};
```

In the public API we expose two functions, which just forward the calls to the fn pointers in the struct:
```C
uint16_t ixy_rx_batch(struct ixy_device* dev, struct pkt_buf* bufs, uint16_t num_bufs);
uint16_t ixy_tx_batch(struct ixy_device* dev, struct pkt_buf* bufs, uint16_t num_bufs);
```

Which of course point to the appropriate driver implementations, so the following is always correct code:
```C
#define IXY_TO_VIRT(dev) container_of(dev, struct virtio_device, ixy)

static uint16_t virtio_rx_batch(struct ixy_device* dev, ...) {
    struct virtio_device* virt = IXY_TO_VIRT(dev);
    ...
}
```
This unwrapping is only needed on the semi-public entry functions to a driver. Internally it can just pass its struct around.

## Pros
 - Unified user-interface, an ixy-fwd app works with every driver
 - Very common practice

## Cons
 - One level of pointer indirection; Performance impact?
 - Too much magic?
 - Even required for just two drivers?

**Comment by emmericp** [comment]
implemented

**Comment by pudelkoM** [comment]
Implemented by #8
